PyMI/src/wmi/__init__.py

```diff
@@ -460,42 +461,54 @@ def __call__(self, timeout_ms=-1):
                            result_code, error_string, error_details):
-        if self._conn_ref:
-            conn = self._conn_ref()
-            if conn:
-                if instance:
-                    event = _Instance(conn,
-                                      instance[u"TargetInstance"].clone(),
-                                      use_conn_weak_ref=True)
-                    try:
-                        previous_inst = _Instance(
-                            conn, instance[u'PreviousInstance'].clone(),
-                            use_conn_weak_ref=True)
-                        object.__setattr__(event, 'previous', previous_inst)
-                    except (mi.error, AttributeError):
-                        # The 'PreviousInstance' attribute may be missing, for
-                        # example in case of a creation event or simply
-                        # because this field was not requested.
-                        pass
-
-                    self._events_queue.append(event)
-                if error_details:
-                    self._error = (
-                        result_code, error_string,
-                        _Instance(
-                            conn, error_details.clone(),
-                            use_conn_weak_ref=True))
-                self._event.set()
+        if not more_results:
+            self._operation_finished.set()
+
+        if instance:
+            event = _Instance(self._conn,
+                              instance[u"TargetInstance"].clone(),
+                              use_conn_weak_ref=True)
+            try:
+                previous_inst = _Instance(
+                    self._conn, instance[u'PreviousInstance'].clone(),
+                    use_conn_weak_ref=True)
+                object.__setattr__(event, 'previous', previous_inst)
+            except (mi.error, AttributeError):
+                # The 'PreviousInstance' attribute may be missing, for
+                # example in case of a creation event or simply
+                # because this field was not requested.
+                pass
+
+            self._events_queue.append(event)
+        if error_details:
+            self._error = (
+                result_code, error_string,
+                _Instance(
+                    self._conn, error_details.clone(),
+                    use_conn_weak_ref=True))
+        self._event.set()
+
+    @avoid_blocking_call
+    def _wait_for_operation_cancel(self):
+        self._operation_finished.wait()
 
     def close(self):
         if self._operation:
             self._operation.cancel()
+            # Those operations are asynchronous. We'll need to wait for the
+            # subscription to be canceled before deallocating objects,
+            # otherwise MI can crash when receiving further events. We rely on
+            # the fact that an event will be emitted once the subscription is
+            # canceled.
+            self._wait_for_operation_cancel()
```

Comment:
hm, So, this will wait indefinetely, until the ``self._operation_finished`` Event is generated. But what if it's not? ``self._operation.cancel()`` shoudl probably generate the Event, but I don't think it does.